In [3]:
from sklearn.datasets import load_digits

In [4]:
digits = load_digits()

In [5]:
digits

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [6]:
digits['data']    #independent variabe

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [7]:
digits.target     #dependent variable i.e., pixel

array([0, 1, 2, ..., 8, 9, 8])

In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

In [9]:
x = digits.data
y = digits.target

In [10]:
x

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [11]:
y

array([0, 1, 2, ..., 8, 9, 8])

In [12]:
x_train, x_test, y_train, y_test=train_test_split(x,y, test_size=0.2, random_state=1)

In [13]:
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(1437, 64) (360, 64) (1437,) (360,)


Build Perceptron Model


In [17]:
from sklearn.linear_model import Perceptron

In [18]:
perceptron = Perceptron()
perceptron.fit(x_train, y_train)

Perceptron()

In [20]:
y_pred_train = perceptron.predict(x_train)
y_pred_test = perceptron.predict(x_test)

In [21]:
confusion_matrix(y_test, y_pred_test)

array([[41,  0,  0,  0,  2,  0,  0,  0,  0,  0],
       [ 0, 34,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 0,  1, 35,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 40,  0,  1,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 38,  0,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0, 29,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 37,  0,  0,  0],
       [ 0,  0,  0,  0,  1,  1,  0, 35,  0,  0],
       [ 0,  0,  0,  0,  0,  2,  0,  0, 27,  0],
       [ 1,  0,  0,  0,  1,  1,  0,  0,  2, 29]])

In [22]:
print(classification_report(y_test, y_pred_test))

              precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.94      0.97      0.96        35
           2       1.00      0.97      0.99        36
           3       1.00      0.98      0.99        41
           4       0.88      1.00      0.94        38
           5       0.85      0.97      0.91        30
           6       1.00      1.00      1.00        37
           7       1.00      0.95      0.97        37
           8       0.93      0.93      0.93        29
           9       1.00      0.85      0.92        34

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



In [25]:
print("Training accuracy", accuracy_score(y_train, y_pred_train))
print("*************************"*5)
print("Test accuracy", accuracy_score(y_test, y_pred_test))

Training accuracy 0.9728601252609603
*****************************************************************************************************************************
Test accuracy 0.9583333333333334


Therefore we can solve Multiclass classification problem by n=10(here) number of classes by perceptron model and we are getting training accuracy as 97% and when we are predecting the data we are getting training accuracy as 95%.


Thus we conclude that their is no bias, no variance and perceptron model can handle multiclass classification problem.
